### PR TITLE
Add a json / json5 / hjson parser

### PR DESCRIPTION
**Pull request by cfrantz**
This series of commits adds a json parser to the serde-annotate libary.  The parser uses a PEG-style grammar via the `pest` crate (starting from the pest grammar in the `json5` crate).  The parser defaults to a maximally permissive mode, but can be customized to forbid certain json extensions.

The maximally permissive parser allows:
- Comments in the 3 common styles (C-like block comments, C++-like slash-slash comments and script-like hash comments).
- Commas in aggregates are optional (hjson-like).
- Aggregates may have a trailing comma (json5-like).
- Integers are permitted in the common bases using the common prefixes (`0b`, `0o`, `0x`).
- Json5 numeric extensions are supported (being lax with decimal-point placement).
- Multiline string blocks are permitted in hjson (triple-quote) and json5 (C-like line continuation) style.
- Unquoted strings are supported in both json5 and hjson style.

This PR adds support for parsing json-like text into a `Document` strucure.

### CODE REVIEW

**Review by jon-flatley — COMMENTED** [review]
Thanks @cfrantz, this is really cool stuff! I don't have too much to comment on, just a couple of minor things. I find the parsing and emitting logic is quite complex and hard to follow, but your included tests seem to cover those areas quite nicely :+1:

File: src/relax.rs [hunk 1]
```diff
@@ -0,0 +1,1084 @@
...
+        if t.starts_with("0x") || t.starts_with("0X") {
+            // Hexadecimal integer.
+            Self::syntax_error(
+                !self.number_hex,
+                "hexadecimal literal",
+                pair.as_span().start_pos(),
+            )?;
+            return Self::from_str_radix(&t, 16, negative);
+        } else if t.starts_with("0b") || t.starts_with("0B") {
+            // Hexadecimal integer.
```
[the remark this answers is not in the document]
Comment (jon-flatley) [reply src/relax.rs:229]: Never mind, I see this is fixed in #2

File: src/relax.rs [hunk 2]
```diff
@@ -0,0 +1,1084 @@
...
+            return Self::from_str_radix(t, 10, negative);
+            /*
+            let val = u128::from_str_radix(t, 10).unwrap();
+            if negative {
+                let val = -(val as i128);
+                return Ok(Document::Int(Int::new(val, Base::Dec)));
+            } else {
+                return Ok(Document::Int(Int::new(val, Base::Dec)));
+            }
+            */
```
[the remark this answers is not in the document]
Comment (cfrantz) [reply src/relax.rs:268]: No, its not needed.  It's also fixed by some `Int` cleanups in #2.

File: src/relax.rs [hunk 3]
```diff
@@ -0,0 +1,1084 @@
...
+        }
+    }
+
+    fn handle_kvpair(&self, pairs: &mut Pairs<Rule>) -> Result<(Document, bool), Error> {
+        let mut k = usize::MAX;
+        let mut v = usize::MAX;
+        let mut kv = vec![];
+        let mut comma = false;
+        loop {
```
[the remark this answers is not in the document]
Comment (cfrantz) [reply src/relax.rs:277]: Both of these loops are using `peek` to look at the next parsed rule to decide whether or not to consume the rule on the current iteration of the loop (e.g. current kvpair) or the next iteration of the loop.  This is done to heuristically guess to which element in the aggregate the comments belong.

Because of the peek behavior, the loop cannot be turned into a `for` loop, but it _can_ be simplified into a `while let` loop.

File: src/relax.rs [hunk 4]
```diff
@@ -0,0 +1,1084 @@
...
+                "missing quotes",
+                pair.as_span().start_pos(),
+            )?;
+            Ok(Document::String(s.trim().into(), StrFormat::Unquoted))
+        }
+    }
+
+    fn handle_pair(&self, pair: Pair<Rule>) -> Result<Document, Error> {
+        match pair.as_rule() {
+            //Rule::value => self.handle_pair(pair.into_inner().next().unwrap()),
```
[the remark this answers is not in the document]
Comment (cfrantz) [reply src/relax.rs:525]: Nope; removed.
